diagnostic_aggregator/src/aggregator.cpp

```diff
@@ -107,13 +106,77 @@ void Aggregator::diagCallback(const diagnostic_msgs::DiagnosticArray::ConstPtr&
   }
 }
 
-Aggregator::~Aggregator() 
+Aggregator::~Aggregator()
 {
   if (analyzer_group_) delete analyzer_group_;
 
   if (other_analyzer_) delete other_analyzer_;
 }
 
+
+void Aggregator::bondBroken()
+{
+  boost::mutex::scoped_lock lock(mutex_); // Possibility of multiple bonds breaking at once
+
+  vector<boost::shared_ptr<bond::Bond> >::iterator it;
+  for (it = bonds_.begin(); it != bonds_.end();)
+  {
+    if ((*it)->isBroken())
+    {
+      string id = (*it)->getId();
+      ROS_DEBUG("bond for namespace %s was broken", id.c_str());
+      analyzer_group_->removeAnalyzer(added_analyzers_.find(id)->second);
+      added_analyzers_.erase(id);
+      bonds_.erase(it); // it points to next in list (or end())
+    } else {
+      // increment in afterthought will go past bonds_.end() if erase()ing last
+      // element, so do it here instead
+      ++it;
+    }
+  }
+}
+
+void Aggregator::bondFormed(){}
+
+bool Aggregator::addDiagnostics(diagnostic_msgs::AddDiagnostics::Request &req,
+				diagnostic_msgs::AddDiagnostics::Response &res)
+{
+  // Wait for connections on this bond. Wait until it's formed to put it in the
+  // bond vector
```

Comment:
In short, it looks like we have two approaches:

1. report the result of adding analysers on the good faith that the bond will form
2. report success on the good faith that the analysers will be successfully added when the bond forms

Both have a chance of leaving the system in an imperfect state.

How about:

**Service Callback**

Set up a bond formed callback with partial binding (below).

Check the validity with `group->init` and return that result, but refrain from calling `analyzer_group->AddAnalyzer` and `added_analyzers_.insert()`. Store the group in an `analyzers_to_be_added_` map (or similar).

**Bond Formed Callback**

Add an argument to this function so you can pass request specific information to the callback (loading namespace, or even the validated group itself). Then when you set it as a callback for the bond, do *partial binding*. Finally, add it to the analysers.